hudi-common/src/main/java/org/apache/hudi/io/storage/HoodieAvroHFileReader.java

```diff
@@ -7,205 +7,136 @@
   private static final Logger LOG = LoggerFactory.getLogger(HoodieAvroHFileReader.class);
 
-  private final Path path;
-  private final FileSystem fs;
-  private final Configuration hadoopConf;
-  private final CacheConfig config;
-  private final Option<byte[]> content;
+  private final Configuration conf;
+  private final Option<Path> path;
+  private final Option<byte[]> bytesContent;
+  private Option<HFileReader> sharedHFileReader;
   private final Lazy<Schema> schema;
 
-  // NOTE: Reader is ONLY THREAD-SAFE for {@code Scanner} operating in Positional Read ("pread")
-  //       mode (ie created w/ "pread = true")
-  // Common reader is not used for the iterators since they can be closed independently.
-  // Use {@link getSharedReader()} instead of accessing directly.
-  private Option<HFile.Reader> sharedReader;
-  // NOTE: Scanner caches read blocks, therefore it's important to re-use scanner
-  //       wherever possible
-  private Option<HFileScanner> sharedScanner;
-
-  private final Object sharedLock = new Object();
-
-  public HoodieAvroHFileReader(Configuration hadoopConf, Path path, CacheConfig cacheConfig) throws IOException {
-    this(path, FSUtils.getFs(path.toString(), hadoopConf), hadoopConf, cacheConfig, Option.empty());
-  }
-
-  public HoodieAvroHFileReader(Configuration hadoopConf, Path path, CacheConfig cacheConfig, FileSystem fs, Option<Schema> schemaOpt) throws IOException {
-    this(path, fs, hadoopConf, cacheConfig, schemaOpt);
-  }
-
-  public HoodieAvroHFileReader(Configuration hadoopConf, Path path, CacheConfig cacheConfig, FileSystem fs, byte[] content, Option<Schema> schemaOpt) throws IOException {
-    this(path, fs, hadoopConf, cacheConfig, schemaOpt, Option.of(content));
-  }
-
-  public HoodieAvroHFileReader(Path path, FileSystem fs, Configuration hadoopConf, CacheConfig config, Option<Schema> schemaOpt) throws IOException {
-    this(path, fs, hadoopConf, config, schemaOpt, Option.empty());
+  public HoodieAvroHFileReader(Configuration conf, Path path, Option<Schema> schemaOption) {
+    this.conf = conf;
+    this.path = Option.of(path);
+    this.bytesContent = Option.empty();
+    this.sharedHFileReader = Option.empty();
+    this.schema = schemaOption.map(Lazy::eagerly)
+        .orElseGet(() -> Lazy.lazily(() -> fetchSchema(getSharedHFileReader())));
   }
 
-  public HoodieAvroHFileReader(Path path, FileSystem fs, Configuration hadoopConf, CacheConfig config, Option<Schema> schemaOpt, Option<byte[]> content) throws IOException {
-    this.path = path;
-    this.fs = fs;
-    this.hadoopConf = hadoopConf;
-    this.config = config;
-    this.content = content;
-
-    // Shared reader is instantiated lazily.
-    this.sharedReader = Option.empty();
-    this.sharedScanner = Option.empty();
-    this.schema = schemaOpt.map(Lazy::eagerly)
+  public HoodieAvroHFileReader(Configuration conf, byte[] content, Option<Schema> schemaOption) {
```

Comment:
do you think we should name this HoodieNativeAvroHfileReader ?

Reply:
Makes sense.  Renamed.